amethyst_assets/src/storage.rs

```diff
@@ -0,0 +1,214 @@
+
+use crossbeam::sync::MsQueue;
+use hibitset::BitSet;
+use specs::{Component, DenseVecStorage, Fetch, FetchMut, System, UnprotectedStorage, VecStorage};
+use specs::common::Errors;
+
+use BoxedErr;
+use asset::Asset;
+use error::AssetError;
+
+/// An `Allocator`, holding a counter for producing unique IDs.
+#[derive(Debug, Default)]
+pub struct Allocator {
+    store_count: AtomicUsize,
+}
+
+impl Allocator {
+    /// Produces a new id.
+    pub fn next_id(&self) -> usize {
+        self.store_count.fetch_add(1, Ordering::Relaxed)
+    }
+}
+
+
+/// An asset storage, storing the actual assets and allocating
+/// handles to them.
+pub struct AssetStorage<A: Asset> {
+    assets: VecStorage<A>,
+    bitset: BitSet,
+    handles: Vec<Handle<A>>,
+    handle_alloc: Allocator,
+    pub(crate) processed: Arc<MsQueue<Processed<A>>>,
+    unused_handles: MsQueue<Handle<A>>,
+}
+
+impl<A: Asset> AssetStorage<A> {
+    /// Creates a new asset storage.
+    pub fn new() -> Self {
+        AssetStorage {
+            assets: Default::default(),
+            bitset: Default::default(),
+            handles: Default::default(),
+            handle_alloc: Default::default(),
+            //new_handles: MsQueue::new(),
+            processed: Arc::new(MsQueue::new()),
+            unused_handles: MsQueue::new(),
+        }
+    }
+
+    /// Allocate a new handle.
+    pub(crate) fn allocate(&self) -> Handle<A> {
+        self.unused_handles
+            .try_pop()
+            .unwrap_or_else(|| self.allocate_new())
+    }
+
+    fn allocate_new(&self) -> Handle<A> {
+        let id = self.handle_alloc.next_id() as u32;
+        let handle = Handle {
+            id: Arc::new(id),
+            marker: PhantomData,
+        };
+
+        handle
+    }
+
+    /// Get an asset from a given asset handle.
+    pub fn get(&self, handle: &Handle<A>) -> Option<&A> {
+        if self.bitset.contains(handle.id()) {
+            Some(unsafe { self.assets.get(handle.id()) })
+        } else {
+            None
+        }
+    }
+
+    /// Get an asset mutably from a given asset handle.
+    pub fn get_mut(&mut self, handle: &Handle<A>) -> Option<&mut A> {
+        if self.bitset.contains(handle.id()) {
+            Some(unsafe { self.assets.get_mut(handle.id()) })
+        } else {
+            None
+        }
+    }
+
+    /// Process finished asset data and maintain the storage.
+    pub fn process<F>(&mut self, mut f: F, errors: &Errors)
+    where
+        F: FnMut(A::Data) -> Result<A, BoxedErr>,
+    {
+        while let Some(processed) = self.processed.try_pop() {
+            let Processed {
+                data,
+                format,
+                handle,
+                name,
+            } = processed;
+            let assets = &mut self.assets;
+            let bitset = &mut self.bitset;
+            let handles = &mut self.handles;
+            errors.execute::<AssetError, _>(|| {
+                let asset = data.and_then(|d| f(d))
```

Comment:
Maybe `and_then(f)`?

Reply:
Tried, doesn't work since it would move the closure.

Reply:
`and_then(&mut f)` in this case

Reply:
Is `FnMut` implemented for `&T where T: FnMut`? I think not, will try later.